controllers/capabilities/tenant_controller.go

```diff
@@ -95,119 +107,116 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
-					return ctrl.Result{}, err
-				}
-			} else {
-				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion", "tenantID", existingTenant.Signup.Account.ID)
-			}
+		// On wait error, retry
+		if helper.IsWaitError(reconcileErr) {
+			reqLogger.Info("ERROR", "wait error", reconcileErr)
+			return ctrl.Result{Requeue: true}, nil
 		}
 
-		controllerutil.RemoveFinalizer(tenantR, tenantFinalizer)
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
+		// On orphan error, do not retry
+		if helper.IsOrphanSpecError(reconcileErr) {
+			reqLogger.Info("ERROR", "orphan error", reconcileErr)
+			return ctrl.Result{Requeue: false}, nil
 		}
 
-		return ctrl.Result{}, nil
+		// On all other errors, retry
+		reqLogger.Error(reconcileErr, "failed to reconcile")
+		r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "ReconcileError", "%v", reconcileErr)
+		return ctrl.Result{}, reconcileErr
 	}
 
-	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
-	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
-	if tenantR.GetDeletionTimestamp() != nil {
-		return ctrl.Result{}, nil
-	}
+	reqLogger.Info("Tenant reconciled successfully")
+	return ctrl.Result{}, nil
+}
 
-	// If the tenant.Status.TenantID is found and the annotation is not found - create
-	// If the tenant.Status.TenantID is found and the annotation is found but, the value of annotation is different to the status.TenantID - update
-	tenantIdAnnotationFound := true
-	tenantId := tenantR.Status.TenantId
-	if value, found := tenantR.ObjectMeta.Annotations[tenantIdAnnotation]; tenantId != 0 && !found || tenantId != 0 && found && value != strconv.FormatInt(tenantR.Status.TenantId, 10) {
-		if tenantR.ObjectMeta.Annotations == nil {
-			tenantR.ObjectMeta.Annotations = make(map[string]string)
-		}
+func (r *TenantReconciler) specReconciler(tenantCR *capabilitiesv1alpha1.Tenant, logger logr.Logger) (*TenantStatusReconciler, error) {
+	// Save a copy of the original tenant CR prior to changing it for status reconciler
+	originalTenantCR := tenantCR.DeepCopy()
 
-		tenantR.ObjectMeta.Annotations[tenantIdAnnotation] = strconv.FormatInt(tenantR.Status.TenantId, 10)
-		tenantIdAnnotationFound = false
+	// Setup porta client
+	portaClient, err := r.setupPortaClient(tenantCR, originalTenantCR, logger)
+	if err != nil {
+		statusReconciler := NewTenantStatusReconciler(r.BaseReconciler, tenantCR, originalTenantCR, err)
+		return statusReconciler, err
 	}
 
-	tenantFinalizerFound := true
-	if !controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
-		controllerutil.AddFinalizer(tenantR, tenantFinalizer)
-		tenantFinalizerFound = false
+	// Process tenant deletion
+	tenantInDeletion, err := r.reconcileDeletion(tenantCR, portaClient, logger)
+	if err != nil || tenantInDeletion {
```

Comment:
I wonder if a tenant can be in deletion state for days, weeks or even months. If that is the case, it does not make sense to have the operator trying again and again.

Reply:
Yes it can, currently, if tenant is marked for deletion already - we do not retrigger the reconciler on the CR because the current error type is isOrphaned.
As noted on following comments the improvements we can make are to remove the isOrphaned error and simply, remove the CR.